blog/content/post/2020-08-05-PCIe-features.md

```diff
@@ -0,0 +1,213 @@
+assigned to the Function and pointer to first capability in linked list.
+
+- **Linked list of capabilities**  
+Each capability has its own header, it contains ID of the
+capability, its version and pointer to the next capability in the list. Beneath
+this header are mapped registers needed by the capability.
+
+To check if a given capability is implemented by Function, a software has to
+search through the list and check if a given capability ID is present in it.
+
+## What PCIe capability actually is
+
+It is nothing more than a predefined feature of the Device Function, a feature
+that is known to be possible to be implemented in all Function, but most of them
+are optional. Every capability has registers in a Device mapped to configuration
+space. Those registers are a kind of interface to the capability. Some are read
+only and contain information needed by the software, others are read/write and
+can be used to pass information to the Function.
+
+## List of capabilities structures:
+(This list can be incomplete)
+
+- **PCI Express Capability register block**
+
+This is one of the most important capability structure, it must be present in
+all PCIe Functions. It is a collection of various information about:
+
+1. **Device** e.g.:  
+
+    - Maximum payload of Transaction Layer Packet size that the Function can support,
+    - Maximum accepted change latency between power states.  
+
+2. **Link** e.g.:  
+
+    - Link max speed \[GT/s\]
+    - Link max width (number of lines)
+    - Link change latency between power states
+    - Port number for given Link  
+
+3. **Port** e.g.:
+  
+    - Physical slot number which is a chassis unique identifier for a slot.  
+
+4. **Hot-Plug**
+
+    Registers responsible for this capability are located in the Capability register
+    block. This capability allows change of Device in the PCIe slot at runtime.
+    By writing to registers corresponding to this capability OS can let the Root or
+    Switch Ports know to power Off or power On. After power is turned Off an user
+    can safely remove the Device, after inserting new one power should be turned On
+    to its Port.
+
+5. **Baseline Error Reporting**
+
+    Bits corresponding to this capability are located partially in the configuration
+    header, partially in the Capability register block. All of them must be present in
+    all PCIe Functions. Some bits of this capability are used to set error reporting,
+    others store status of errors.
+
+- **Power Management**
+
+OS can manage the power environment of a Function directly by accessing registers
+corresponding to this capability. OS can set Device state to one of 4 states. 2
+of them (D0, D4) are mandatory and other 2 (D2, D3) can be optionally implemented.
+D0 state is a state where the Function is fully operational and uses full power,
+every next state (D1,D2,D3) uses lower power, but also takes more time to recover
```

Comment:
Here are D0, D1, D2 and D3, while the list above doesn't have D1 but has D4 instead. One of those is probably wrong

Reply:
you right. there are 4 states: D0 D1 D2 D3.